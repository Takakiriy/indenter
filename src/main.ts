import * as clipboardy from 'clipboardy';
import { pp } from "./lib";

// main
export async function  main() {

    const  inputText = clipboardy.readSync();
    const  lines = inputText.split('\n');
    const  nullLength = 9999;
    const  indentRegularExpression = /^( |Â¥t)*/;
    if (inputText.trim() === '') {
        return;
    }
    var  minIndentLength = nullLength;
    for (const line of lines) {
        if (line.trim() !== '') {
            const  indentLength = indentRegularExpression.exec( line )![0].length;
            minIndentLength = Math.min(minIndentLength, indentLength)
        }
    }
    const  firstLineIndent = indentRegularExpression.exec( lines[0] )![0];
    if (minIndentLength === nullLength) {
        return;
    }

    if (minIndentLength < firstLineIndent.length) {

        // insert indent
        for (var lineNum = 2;  lineNum < lines.length;  lineNum += 1 ) {
            if (lines[lineNum - 1] !== '') {

                lines[lineNum - 1] = firstLineIndent + lines[lineNum - 1];
            }
        }
    } else {

        // cut indent
        for (var lineNum = 1;  lineNum < lines.length;  lineNum += 1 ) {
            lines[lineNum - 1] = lines[lineNum - 1].substr(minIndentLength);
        }
    }

    const  outputText = lines.join('\n');
    clipboardy.writeSync(outputText);
}


// getStdOut
// Example:
//    var d = getStdOut();  // Set break point here and watch the variable d
function  getStdOut(): string[] {
    return  stdout.split('\n');
}

// println
// #keyword: println, console.log, consoleLog
// Output any text to standard output.
export function  println(message: any, delayedExpanding: boolean = false) {
    if (typeof message === 'object' && !delayedExpanding) {
        message = JSON.stringify(message);
    }
    if (withJest && !delayedExpanding) {
        stdout += message.toString() + '\n';
        pp(message.toString());
    } else {
        consoleLog(message);
    }
}
const  consoleLog = console.log;
console.log = println;

// callMainFromJest
// #keyword: callMainFromJest
export async function  callMainFromJest(parameters?: string[], options?: {[name: string]: string}) {
    withJest = true;
    stdout = '';
    if (parameters) {
        programArguments = parameters;
    } else {
        programArguments = [];
    }
    if (options) {
        programOptions = options;
    } else {
        programOptions = {};
    }

    await main();
}

var    locale = '';
var    withJest = false;
export var  stdout = '';
export var  programArguments: string[] = [];
export var  programOptions: {[key: string]: any} = {};
