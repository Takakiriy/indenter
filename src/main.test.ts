import * as main from './main';
import * as path from "path";
import * as clipboardy from 'clipboardy';
import * as lib from "./lib";
import * as fs from "fs";
const  callMain = main.callMainFromJest;

if (path.basename(process.cwd()) !== 'src') {
    // Because the second execute of Jest watch mode is inherited the current folder.
    process.chdir('src');
}

test.each([
    ["insert indent", []],
    ["insert line 2 indent", []],
    ["cut indent", []],
    ["cut indent at the last line", []],
    ["change tab size 2 to 4 and insert indent", ["24"]],
    ["change tab size 4 to 2 and cut indent", ["42"]],
    ["change YAML tab size 2 to 4 and insert indent", ["24"]],
    ["change YAML tab size 4 to 2 and cut indent", ["42"]],
    ["change 2 spaces to tab and cut indent", ["2t"]],
    ["change 4 spaces to tab and insert indent", ["4t"]],
    ["change tab to 2 spaces and cut indent", ["t2"]],
    ["change tab to 4 spaces and insert indent", ["t4"]],
])("%s", async (caseName, parameters) => {
    const  inputText = lib.getSnapshot(`${caseName}: 1 sourceFileContents 1`);
    clipboardy.writeSync(inputText);

    await callMain(parameters, {});
    const  outputText = clipboardy.readSync();
    expect(outputText).toMatchSnapshot('answer');
});

describe("test of test >>", () => {
    test("checks snapshots files are confirmed", () => {
        const  activeSnapshots = fs.readFileSync('__snapshots__/main.test.ts.snap').toString();
        const  backUpSnapshots = fs.readFileSync('__snapshots__/main.test.ts.snap.confirmed-ts').toString();
            // 拡張子の末尾を .snap にしない理由は、Jest が使っていない .snap ファイルを自動的に削除しようとするからです
            // ____.snap.confirmed-ts ファイルが存在する理由は、Jest の自動編集が予期しないデータを追加することがあるからです

        expect(activeSnapshots).toBe(backUpSnapshots);
    });
});
