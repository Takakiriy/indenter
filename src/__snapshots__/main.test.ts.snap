// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change 2 spaces to tab and cut indent: 1 sourceFileContents 1`] = `
"      Section() {
        Section:

          SubSection:
      }
"
`;

exports[`change 2 spaces to tab and cut indent: answer 1`] = `
"Section() {
	Section:

		SubSection:
}
"
`;

exports[`change 4 spaces to tab and insert indent: 1 sourceFileContents 1`] = `
"        Section() {
    Section:

        SubSection:
}
"
`;

exports[`change 4 spaces to tab and insert indent: answer 1`] = `
"		Section() {
			Section:

				SubSection:
		}
"
`;

exports[`change tab size 2 to 4 and insert indent: 1 sourceFileContents 1`] = `
"        Section() {
  Section:

    SubSection:
}
"
`;

exports[`change tab size 2 to 4 and insert indent: answer 1`] = `
"        Section() {
            Section:

                SubSection:
        }
"
`;

exports[`change tab size 4 to 2 and cut indent: 1 sourceFileContents 1`] = `
"        Section() {
            Section:

                SubSection:
        }
"
`;

exports[`change tab size 4 to 2 and cut indent: answer 1`] = `
"Section() {
  Section:

    SubSection:
}
"
`;

exports[`change tab to 2 spaces and cut indent: 1 sourceFileContents 1`] = `
"			Section() {
				Section:

					SubSection:
			}
"
`;

exports[`change tab to 2 spaces and cut indent: answer 1`] = `
"Section() {
  Section:

    SubSection:
}
"
`;

exports[`change tab to 4 spaces and insert indent: 1 sourceFileContents 1`] = `
"			Section() {
	Section:

		SubSection:
}
"
`;

exports[`change tab to 4 spaces and insert indent: answer 1`] = `
"            Section() {
                Section:

                    SubSection:
            }
"
`;

exports[`cut indent: 1 sourceFileContents 1`] = `
"        Section 1:
        Section 2:
          Section 2-1:

            Section 2-1-1:
        Section 3:
"
`;

exports[`cut indent: answer 1`] = `
"Section 1:
Section 2:
  Section 2-1:

    Section 2-1-1:
Section 3:
"
`;

exports[`insert indent: 1 sourceFileContents 1`] = `
"        Section 1:
Section 2:
  Section 2-1:

    Section 2-1-1:
Section 3:
"
`;

exports[`insert indent: answer 1`] = `
"        Section 1:
        Section 2:
          Section 2-1:

            Section 2-1-1:
        Section 3:
"
`;

exports[`insert line 2 indent: 1 sourceFileContents 1`] = `
"        Section 1:
  Section 1-1:

    Section 1-1-1:
"
`;

exports[`insert line 2 indent: answer 1`] = `
"        Section 1:
          Section 1-1:

            Section 1-1-1:
"
`;
